### DISCUSSION TITLE
asn1_d2i_read_bio() fails on short read

### DISCUSSION THREAD

**Discussion by arnoldgeels** [openssl/openssl#22484] [Q&A]
I believe I have found a bug in asn1_d2i_read_bio() when a read is shorter than expected.  

I found this when I was reading a DER-encoded message from a socket and the server happened to send a packet of four bytes.
The call fails with `error:0680007B:asn1 encoding routines::header too long`.

Reproducing such a short read from a socket depends on timing and is not reliable, so I found a way to reproduce the issue with just a memory BIO and a creative callback.  Since the actual DER-encoded message doesn't matter, I use a certificate as the message.  The code below assumes a file with a certificate is present.  I tested with the certificate from www.openssl.org.

I found this issue in OpenSSL 3.0.9 on Win64 and I am also able to reproduce this in master.

Let me know if you agree this is a bug.  I was able to work around this issue using a another creative BIO callback, so this is not urgent.

```
#include <stdio.h>
#include <openssl/bio.h>
#include <openssl/err.h>
#include <openssl/x509.h>
#ifdef _WIN32
# include <openssl/applink.c>
#endif

#define SHORT_READ 4

/*-
 * Every time a read is about to be attempted, write a small portion 
 * into the mem bio. len may be less than SHORT_READ but that just means 
 * it will not read the whole portion.
 */
long shortReadCallback(BIO* b, int oper, const char* argp, size_t len, int argi,
                       long argl, int ret, size_t* processed)
{
    if (oper == BIO_CB_READ) {
        size_t readbytes;
        char buf[SHORT_READ];
        BIO* arg = (BIO*)BIO_get_callback_arg(b);
        if (BIO_read_ex(arg, buf, SHORT_READ, &readbytes) != 1 ||
                        BIO_write_ex(b, buf, readbytes, NULL) < 0)
            ERR_print_errors_fp(stdout);
    }
    return ret;
}

/*-
 * Supply a DER-encoded certificate in file "www.openssl.org.der".
 * A certificate is used only as example.
 * The issue is generic to all d2i methods.
 */
int main()
{
    BIO* certbio = BIO_new_file("www.openssl.org.der", "rb");
    if (!certbio) {
        puts("failure reading test input");
        puts(ERR_error_string(ERR_get_error(), NULL));
        return 1;
    }
    BIO* membio = BIO_new(BIO_s_mem());
    BIO_set_callback_arg(membio, (char *)certbio);
    BIO_set_callback_ex(membio, shortReadCallback);
    /* fails with ASN1_R_HEADER_TOO_LONG */
    X509* cert = d2i_X509_bio(membio, NULL);
    if (cert) printf("OK\n");
    else puts(ERR_error_string(ERR_get_error(), NULL));
    return !cert;
}
```

**Comment by t8m** [comment]
Quickly looking - in your callback you do not seem to handle the case when `len < SHORT_READ`, could it be a problem?

**Comment by arnoldgeels** [comment]
Well spotted. That would be an issue but it does not happen.  The method starts by reading 8 bytes, gets only 4, and fails.
I have adjusted the test code.  

You can set SHORT_READ to 8 and the issue goes away.

**Comment by t8m** [comment]
This should fix it.

``` diff
diff --git a/crypto/asn1/a_d2i_fp.c b/crypto/asn1/a_d2i_fp.c
index 4af2276a8d..999df25590 100644
--- a/crypto/asn1/a_d2i_fp.c
+++ b/crypto/asn1/a_d2i_fp.c
@@ -148,6 +148,8 @@ int asn1_d2i_read_bio(BIO *in, BUF_MEM **pb)
                     goto err;
                 }
                 len += i;
+                if (i < want)
+                    continue;
             }
         }
         /* else data already loaded */
```

Except the problem is a short read can be also 0 bytes long and in that case we might want to also continue if 0 was returned. Except 0 is returned also on EOF so we would have to check for BIO_eof() before continuing. And the remaining question is, what other use-cases this will break. I have a strong suspicion that supporting short reads here could have unwanted consequences for other existing (albeit perhaps exotic) use-cases.

I would recommend adding some kind of buffering - i.e. read the whole data into a buffer and then use a d2i function on that buffer.

**Comment by t8m** [comment]
More complete patch:

```diff
diff --git a/crypto/asn1/a_d2i_fp.c b/crypto/asn1/a_d2i_fp.c
index 4af2276a8d..97197ec508 100644
--- a/crypto/asn1/a_d2i_fp.c
+++ b/crypto/asn1/a_d2i_fp.c
@@ -142,12 +142,14 @@ int asn1_d2i_read_bio(BIO *in, BUF_MEM **pb)
                 ERR_raise(ERR_LIB_ASN1, ASN1_R_NOT_ENOUGH_DATA);
                 goto err;
             }
-            if (i > 0) {
+            if (i >= 0) {
                 if (len + i < len) {
                     ERR_raise(ERR_LIB_ASN1, ASN1_R_TOO_LONG);
                     goto err;
                 }
                 len += i;
+                if (i < want && !BIO_eof(in))
+                    continue;
             }
         }
         /* else data already loaded */
```

**Comment by arnoldgeels** [comment]
Thank you for the quick response.

I have applied your patch into my master build.  The test program still fails, but that is probably because my "creative callback" does not intercept the call to `BIO_eof()`.  I will apply the patch to 3.0.9 and use it in the program that had the original issue with the socket.

> I have a strong suspicion that supporting short reads here could have unwanted consequences for other existing (albeit perhaps exotic) use-cases.

I agree, that is a risk.  This "creative callback" is an actually an example.

Adding a buffer: I believe the code correctly handles short reads for the contents.  The issue is "just" for reading the header.

**Comment by arnoldgeels** [comment]
I have tested the patch in the program that had the original issue with the socket. Of course I took out my workaround. Even if the first read is short this is now handled correctly, the error "header too long" no longer appears.  Nice work!

Are you going to apply the patch?  Or are you still thinking about adding some kind of buffering?

**Comment by t8m** [comment]
I'll submit the patch as a PR - I am not sure it should be merged though as it might have unwanted consequences.

**Comment by davidben** [comment]
> Except the problem is a short read can be also 0 bytes long and in that case we might want to also continue if 0 was returned. Except 0 is returned also on EOF so we would have to check for BIO_eof() before continuing.

This doesn't make sense. In a `read`-style API, zero is unambiguously EOF. The reason to return a short read is if you have some data available on the stream, but not the full buffer. You don't want to block the call (perhaps that data is never coming), so you immediately return.

If you don't have any data available on the stream, you would either block, or fail (-1) and surface some retry signal, depending on whether the stream is blocking or non-blocking. The non-blocking case can be ignored because the d2i functions cannot support non-blocking `BIO`s anyway. (There's nowhere to stash the partial state and signal the caller to retry the operation.) Thus you should treat zero as EOF and -1 as error.

**Comment by t8m** [comment]
> This doesn't make sense. In a `read`-style API, zero is unambiguously EOF. The reason to return a short read is if you have some data available on the stream, but not the full buffer. You don't want to block the call (perhaps that data is never coming), so you immediately return.

This wouldn't make sense if we could be 100% sure that the BIO follows the semantics you wrote. Also the read could be retryable (i.e., EINTR occured on the blocking call) but the current BIO implementations do not properly support BIO_FLAG_SHOULD_RETRY (except for bio_ssl).

**Comment by obatysh** [comment]
We have encountered the same issue with "error:0680007B:asn1 encoding routines::header too long" error when parsing CMS structure for enveloped data that uses streaming (contains a lot of OCTET STRING objects) using d2i_CMS_bio function.
We are using our own implementation of BIO that allows short reads, but situations when very small amount, like less than 5 bytes, is returned are very rare. So the issue happens very sporadically and when ASN1 structure is long and contains a lot of objects (like when streaming is used) chances to encounter it are much higher.
I checked the proposed patch and it fixes the problem.

**Comment by t8m** [comment]
https://github.com/openssl/openssl/pull/22486 I've submitted a PR with the first proposed solution as it should be safe. The second solution could potentially cause endless loops with incorrect BIOs.
